File: app/main-api.py
import os
from fastapi.responses import HTMLResponse, FileResponse, StreamingResponse
from fastapi import FastAPI, File, UploadFile, Request
import uvicorn
from answer import answer
from pydantic import BaseModel
from typing import Any, Union
import traceback
import threading
import string
import random
from pdf import pdf_to_word
import logging
import time
from fastapi.middleware.cors import CORSMiddleware

logging.basicConfig(level=logging.INFO)

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins= ["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)


class TextItem(BaseModel):
    text: str = "旅行者们，你好啊"

import uuid


@app.post("/text-to-audio")
async def http_text_to_audio(
    text_item: TextItem,
):

    wav_file_path = "/tmp/test.wav"
    try:
        from  vits.vits import text_to_audio_file
        wav_file_path = text_to_audio_file(text_item.text)

        respose = FileResponse(path=wav_file_path, filename= os.path.basename(wav_file_path))
        return respose
    finally:
        async def clean():
            await asyncio.sleep(10)
            if(os.path.exists(wav_file_path)):
                os.remove(wav_file_path)
        asyncio.create_task(clean())

@app.post("/pdf-to-word")
async def http_pdf_to_word(
    file: UploadFile = File(description="file as UploadFile"),
):
    word_file_path = "/tmp/test.docx"

    try:
        pdf_file_name = file.filename

        if not pdf_file_name.endswith('.pdf'):
            return "请上传 pdf 文件"

        logging.info(f"upload filename: {pdf_file_name} !")

        temp_dir = "/tmp"

        file_name = os.path.basename(pdf_file_name)

        random_s = "pdf-to-word-" + uuid.uuid4().hex

        pdf_file_path = os.path.join(temp_dir, random_s + pdf_file_name)
        word_file_name = file_name + '.docx'
        word_file_path = os.path.join(temp_dir, random_s + word_file_name)

        content = file.file.read()
        with open(pdf_file_path, 'wb') as f:
            f.write(content)

        pdf_to_word(pdf_file_path, word_file_path)

        respose = FileResponse(path=word_file_path, filename=word_file_name)

        os.remove(pdf_file_path)

        return respose
    finally:
        async def clean():
            await asyncio.sleep(10)
            if(os.path.exists(word_file_path)):
                os.remove(word_file_path)
        asyncio.create_task(clean())


current_path = os.path.dirname(os.path.abspath(__file__))

chatbot_html_file_path = os.path.join(current_path, "../src/chatbot.html")
pdf_html_file_path = os.path.join(current_path, "../src/pdf.html")
audio_html_file_path = os.path.join(current_path, "../src/text-tu-audio.html")


@app.get("/pdf", response_class=HTMLResponse)
async def index():
    with open(pdf_html_file_path, "r") as f:
        html = f.read()
    return html


@app.get("/text-to-audio", response_class=HTMLResponse)
async def index():
    with open(audio_html_file_path, "r") as f:
        html = f.read()
    return html

class Question(BaseModel):
    text: str = "用户: 写首诗\n小元: "
    # description: Union[str, None] = None
    max_len: Union[float, None] = 50
    temperature: Union[float, None] = 1.0
    top_p: Union[float, None] = 0.95
    top_k: Union[float, None]
    sample: Union[None, bool] = True


@app.post('/api/generate')
async def api_generate(q: Question, request: Request):
    """
    curl -X 'POST' \
    'http://localhost:8000/api/generate' \
    -H 'accept: application/json' \
    -H 'Content-Type: application/json' \
    -d '{
    "text": "用户: 用 markdown 格式写一篇介绍 python 冒泡排序的文章\n小元",
    "max_len": 5000,
    "temperature": 1,
    "top_p": 0.95,
    "sample": true
    }'
    """
    data = await request.json()
    if 'text' not in data or not isinstance(data['text'], str):
        return {
            'ok': False,
            'error': 'Invalid text in post data',
        }
    try:

        generation_params = {
            "text": q.text,
            "max_new_tokens": q.max_len,
            "sample": q.sample,
            "top_k": q.top_k,
            "top_p": q.top_p,
            "request": request,
            "temperature": q.temperature
        }

        ret = answer(**generation_params)

        return {
            'ok': True,
            'text': ret,
        }

    except Exception:
        return {
            'ok': False,
            'error': traceback.format_exc(),
        }


@app.post('/api/generate')
async def api_generate(q: Question, request: Request):
    data = await request.json()
    if 'text' not in data or not isinstance(data['text'], str):
        return {
            'ok': False,
            'error': 'Invalid text in post data',
        }
    try:

        generation_params = {
            "text": q.text,
            "max_new_tokens": q.max_len,
            "sample": q.sample,
            "top_k": q.top_k,
            "top_p": q.top_p,
            "request": request,
            "temperature": q.temperature
        }

        ret = answer(**generation_params)

        return {
            'ok': True,
            'text': ret,
        }

    except Exception:
        return {
            'ok': False,
            'error': traceback.format_exc(),
        }

# from queue import Queue
import asyncio
from concurrent.futures import ThreadPoolExecutor
global_executor = ThreadPoolExecutor(2)

async def generate_text_async(q: Question, request: Request):

    # qu = Queue()
    queue = asyncio.Queue()

    global_loop = asyncio.get_event_loop()

    request.state.is_done = False
    # count = 0

    async def check_is_done():
        while not request.state.is_done:
            await asyncio.sleep(.2)
            is_disconnected = await request.is_disconnected()
            if is_disconnected:
                request.state.is_done = True
                break

    def my_callback(x):
        # nonlocal count

        if request.state.is_done :
            raise Exception("end")

        # queue.put_nowait(x)
        global_loop.call_soon_threadsafe(queue.put_nowait, x)

        # i_next_item = f"count = {count} \t is_done = {request.state.is_done}"
        # print(i_next_item)
        # count += 1
        # asyncio.run(check_is_done)
        # asyncio.run_coroutine_threadsafe(check_is_done)
        # global_loop.run_until_complete(check_is_done)


        # qu.join() # Blocks until task_done is called

    generation_params = {
        "text": q.text,
        "max_new_tokens": q.max_len,
        "sample": q.sample,
        "ondata": my_callback,
        "top_k": q.top_k,
        "top_p": q.top_p,
        # "request": request,
        "temperature": q.temperature
    }


    def task():
        # answer(**generation_params)

        try:
            return answer(**generation_params)
        except:
            pass
        # nonlocal count
        # while True:
        #     time.sleep(1)
        #     generation_params["ondata"](f"data = {count}")

    global_loop.run_in_executor(global_executor, task)

    asyncio.create_task(check_is_done())


    # request.state.is_done = True


    # Consumer
    while True:
        # print(f"is_done = {is_done}")
        next_item = await queue.get() # Blocks until an input is available
        # await asyncio.sleep(1)

        if next_item is None:
            break
        yield next_item
        # queue.task_done() # Unblocks the producer, so a new iteration can start



@app.post("/api/generate_text_async")
async def generate_text(q: Question, request: Request):
    return StreamingResponse(generate_text_async(q, request=request), media_type="text/event-stream")




@app.get("/", response_class=HTMLResponse)
async def index():
    with open(chatbot_html_file_path, "r") as f:
        html = f.read()
    return html


if __name__ == "__main__":
    uvicorn.run("main-api:app", host="0.0.0.0", port=3000)

File: src/main-api.ts
import fastifyFactory, { FastifySchema } from 'fastify';
import fs from 'fs';
import { answer_sync } from './py-utils';
import { pdf_to_word } from './py-utils';
import { fileURLToPath } from 'url';
// import fastifyMultipart from '@fastify/multipart';
// import fastifySee from 'fastify-sse';
import fastifySwagger from '@fastify/swagger';
// import { Fas } from '@fastify/swagger-ui';
import { createReadStream } from 'fs';
// import { PassThrough } from 'stream';
// import { createWriteStream } from 'fs';
// import { pipeline } from 'stream';
// import { createInterface } from 'readline';
import path, { dirname, join } from 'path';
// import { randomBytes } from 'crypto';
import { createRequire } from 'module';
// import dotenv from 'dotenv';

// dotenv.config();

const require = createRequire(import.meta.url);

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

class Question {
  text: string;
  max_len: number;
  temperature: number;
  top_p: number;
  top_k: number;
  sample: boolean;
  constructor({
    text = '用户: 写首诗\n小元: ',
    // description = null,
    max_len = 50,
    temperature = 1.0,
    top_p = 0.95,
    top_k = 0,
    sample = true,
  } = {}) {
    this.text = text;
    // this.description = description;
    this.max_len = max_len;
    this.temperature = temperature;
    this.top_p = top_p;
    this.top_k = top_k;
    this.sample = sample;
  }
}

// Run the server!
const start = async () => {
  const { FastifyFormidable, ajvBinaryFormat } = require('fastify-formidable');

  const app = fastifyFactory({
    logger: true,
    ajv: {
      plugins: [ajvBinaryFormat],
    },
  });

  try {
    // await app.register(fastifyMultipart, { attachFieldsToBody: true });

    await app.register(FastifyFormidable, {
      addContentTypeParser: true,
    });

    // await app.register(fastifySee);

    await app.register(fastifySwagger, {
      openapi: {
        info: {
          title: 'WX API',
          description: 'API for WX',
          version: '3.0.0',
        },
        // consumes: ['multipart/form-data', 'application/json'],
        // produces: ['multipart/form-data', 'application/json'],
        tags: [{ name: 'file', description: 'File operations' }],
        // definitions: {
        //   File: {
        //     type: 'object',
        //     properties: {
        //       file: { type: 'string', format: 'binary' },
        //     },
        //   },
        // },
      },
    });

    const swaggerUi = require('@fastify/swagger-ui');

    await app.register(swaggerUi, {
      routePrefix: '/docs',
      // uiConfig: {
      //   docExpansion: 'full',
      //   deepLinking: false,
      // },
      // uiHooks: {
      //   onRequest: function (request, reply, next) {
      //     next();
      //   },
      //   preHandler: function (request, reply, next) {
      //     next();
      //   },
      // },
      // staticCSP: true,
      // transformStaticCSP: (header) => header,
      // transformSpecification: (swaggerObject, request, reply) => {
      //   return swaggerObject;
      // },
      // transformSpecificationClone: true,
    });

    const random_string = () => {
      return Math.random().toString(16).substring(2, 16);
    };

    app.post(
      '/pdf-to-word',
      {
        schema: {
          tags: ['file'],
          body: {
            type: 'object',
            required: ['file'],
            properties: {
              file: { type: 'string', format: 'binary' },
              // aa: { type: 'string' },
            },
          },
          consumes: ['multipart/form-data'],
          // produces: ['application/json'],
        } as FastifySchema,
      },
      async (request, reply) => {
        try {
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          const { file } = request.body;

          // const pdf_file = file;
          // const pdf_file_name = pdf_file.filename || `pdf-file-${random_string()}.pdf`;

          // // if (!pdf_file_name.endsWith('.pdf')) {
          // //   return '请上传 pdf 文件';
          // // }

          // app.log.info(`upload filename: ${pdf_file_name} !`);

          // const temp_dir = '/tmp';

          // const file_name = path.basename(pdf_file_name);

          // const random_s = 'pdf-to-word-' + random_string();

          // const pdf_file_path = `${temp_dir}/${random_s}${pdf_file_name}`;
          // const word_file_name = `${file_name}.docx`;
          // const word_file_path = `${temp_dir}/${random_s}${word_file_name}`;

          // const content = pdf_file.file;
          // fs.writeFileSync(pdf_file_path, content);

          const pdf_file_path = file;
          const random_s = 'pdf-to-word-' + random_string();
          const file_name = path.basename(pdf_file_path);
          const word_file_name = `${file_name}.docx`;
          const temp_dir = '/tmp';
          const word_file_path = `${temp_dir}/${random_s}${word_file_name}`;

          pdf_to_word(pdf_file_path, word_file_path);

          const stream = fs.createReadStream(word_file_path);
          reply.type(
            'application/vnd.openxmlformats-officedocument.wordprocessingml.document'
          );
          reply.header(
            'Content-Disposition',
            `attachment; filename="${word_file_name}"`
          );
          // eslint-disable-next-line @typescript-eslint/ban-ts-comment
          // @ts-ignore
          // stream.pipe(reply.res);
          stream?.on('end', () => {
            fs.unlinkSync(pdf_file_path);
            fs.unlinkSync(word_file_path);
          });
          return reply.send(stream);
        } catch (err) {
          app.log.error(err);
          reply.code(500).send();
        }
      }
    );

    const chatbot_html_file_path = join(__dirname, 'chatbot.html');

    app.get('/', async (request, reply) => {
      // reply.send({ hello: 'world' });
      // console.log(`file: ${chatbot_html_file_path}`);
      const fileStream = createReadStream(chatbot_html_file_path, 'utf8');
      reply.type('text/html');
      return reply.send(fileStream);
    });

    const generate_text_async = function generate_text_async(
      q
      // request
    ) {
      const generation_params = {
        text: q.text,
        max_new_tokens: q.max_len,
        sample: q.sample,
        top_k: q.top_k,
        top_p: q.top_p,
        // request,
        temperature: q.temperature,
      };
      // yield JSON.stringify(generation_params);

      return answer_sync(generation_params);
    };

    app.post(
      '/api/generate_text_async',
      {
        schema: {
          // tags: ['file'],
          body: {
            type: 'object',
            required: ['text'],
            properties: {
              text: { type: 'string', default: '你好' },
              max_len: { type: 'number', default: 40 },
              sample: { type: 'boolean' },
            },
          },
          consumes: ['application/json'],
          produces: ['text/event-stream'],
        } as FastifySchema,
      },
      async (request, reply) => {
        reply.raw.setHeader('content-type', 'text/event-stream');

        const text_request = request.body;
        const stream = generate_text_async(new Question(text_request as any));

        const breakFn = () => {
          stream.destroy();
          console.log('===========breakFn======');
        };

        request.socket.on('error', breakFn);

        stream.on('data', (chunk) => {
          // console.log({ chunk });
          reply.raw.write(chunk);
        });

        stream.on('end', () => {
          reply.raw.end();
        });

        return reply;

        // const stream = new Readable({
        //   objectMode: true,
        //   read() {
        //     // no thing
        //   },
        // });

        // const delay = (t = 0) => {
        //   return new Promise((re) => {
        //     setTimeout(() => {
        //       re(null);
        //     }, t);
        //   });
        // };

        // (async () => {
        //   let i = 0;
        //   while (i < 10) {
        //     stream.push(`\ndata: ${i}\n`);

        //     await delay(1000);

        //     i++;
        //   }
        //   stream.push('\nend\n');
        //   stream.push(null);
        // })();

        // return reply.send(stream);
      }
    );

    app.post(
      '/api/generate_text',
      {
        schema: {
          // tags: ['file'],
          body: {
            type: 'object',
            required: ['text'],
            properties: {
              text: { type: 'string', default: '你好' },
              max_len: { type: 'number', default: 40 },
              sample: { type: 'boolean' },
            },
          },
          consumes: ['application/json'],
          produces: ['application/json'],
        } as FastifySchema,
      },
      async (request, reply) => {
        const text_request = request.body;
        return reply.send(text_request);
      }
    );

    const res = await app.listen({
      port: 3000,
      host: '0.0.0.0',
    });
    console.log(`Server listening at ${res}`);
  } catch (err) {
    app.log.error(err);
    process.exit(1);
  }
};

start();
